### PR TITLE
default_alloc_error_hook: explain difference to default __rdl_oom in alloc

### PR DESCRIPTION
**Pull request by RalfJung**
Though I'm not sure if that is really the reason that this code is duplicated. On no_std it may already be possible to call user-defined code on allocation failure.

### CODE REVIEW

**Comment by RalfJung** [comment]
Hm, actually there is a subtle difference in behavior... currently the default path taken does *not* invoke any user-defined code. That seems worth preserving for now. I'll change the PR to just add a comment instead.

**Comment by RalfJung** [comment]
r? libs

File: library/std/src/alloc.rs [hunk 1]
```diff
@@ -353,6 +353,9 @@ fn default_alloc_error_hook(layout: Layout) {
     if unsafe { __rust_alloc_error_handler_should_panic != 0 } {
         panic!("memory allocation of {} bytes failed", layout.size());
     } else {
+        // This is the default path taken on OOM, and the only path taken on stable with std.
+        // Crucially, it does *not* call any user-defined code, and therefore there are no potential
+        // reentrancy issues. That makes it different from the default `__rdl_oom` defined in alloc.
```
Comment (workingjubilee) [inline library/std/src/alloc.rs:358]: > no potential reentrancy issues.

innit more like "no potential reentrancy issues aside from the ones we foist on ourselves"?

`__rdl_oom` is an awfully descriptive name, congratulations on us for coming up with that one.

Comment (workingjubilee) [reply library/std/src/alloc.rs:358]: You describe two possible differences, and either of those can independently vary, technically. You should note that `__rdl_oom` calls like `handle_alloc_error()` or whatever it actually does.

Comment (RalfJung) [reply library/std/src/alloc.rs]: I have tried to clarify both of these points, does that help?

Comment (workingjubilee) [reply library/std/src/alloc.rs]: Yes, thank you!

**Pull request merged**